Extract log handlers from WC_Logger

sirreal opened the pull request:
Please continue any discussion on #12716.

~~~Please continue any discussion on #12575.~~~

Implements #10468.

This PR adds extension points for `WC_Logger`. Notably the filter `woocommerce_register_log_handlers` is provided to hook up a log handler. The interface has been modified to loosely adhere to the PHP Standards Recommendations [Logger Interface](http://www.php-fig.org/psr/psr-3/). Specifically, logs are now added via `log`, or one of the level helper methods such as `emergency`.

The abstract class `WC_Log_Handler` has been provided to facilitate implementing new log handlers.

A default log handler `WC_Log_Handler_File` is provided. This handler is attached by default via the provided filter.

These changes attempt to maintain backwards compatibility, `WC_Logger->add()` now adds an `info` level message and has been deprecated.

Implementations for email and DB loggers have been provided, as well as an interface for viewing/clearing DB log messages.

I'd like to get feedback from team members and @gedex. Is this implementation in line with your needs? Does anyone foresee limitations or problems in the direction being taken?

sirreal committed:
Extract file log handler

sirreal committed:
WC_Logger registers and delegates logging to handlers

sirreal committed:
Register file log handler by default

Add wc_register_file_log_handler function.
Calls add_filter with register function to include handler by default

sirreal committed:
Update tests for WC_Logger

sirreal committed:
Add abstract WC_Log_Handler class

Abstract base class is implemented with default formatting implementations.
`$timestamp` has been added to the `WC_Log_Handler::handle` implementation. This ensures all handlers are passed the same timestamp for the same message.
`_legacy` context has been added to `WC_Log_Handler_File` to ensure handle deprecated `WC_Logger::add` provides the same log entries as before.

Fix CS problems.

sirreal committed:
Add logic for log level filtering to `WC_Log_Handler`.

sirreal committed:
Adds tests for bubbling, etc.

Add tests teardown to cleanup log files
Fix errors in log handlers

sirreal committed:
Remove 2nd filter parameter that would never be called

sirreal committed:
Add tag context test for file log handler

sirreal committed:
Use wp_parse_args for default args

sirreal committed:
Cleanup "A" test log file

sirreal committed:
Remove redundant __constructor

sirreal committed:
Add email log handler

sirreal committed:
Extract anonymous functions to improve test compatibility.

sirreal committed:
Provide basic messages for email log handler

sirreal committed:
Fix tests for compatibility with PHP5

sirreal committed:
Restore deprecated functions WC_Logger::clear & WC_Logger::remove

Also updates clear test to supply required argument.

sirreal committed:
Use older test mock syntax for legacy compatibility.

gedex commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,135 @@
+	/**
+	 * Constructor for log handler.
+	 *
+	 * @param arr $args {
```
param type should be array instead of arr?

gedex commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,135 @@
+	public function __construct( $args = array() ) {
+
+		$args = wp_parse_args( $args, array(
+			'threshold' => WC_Logger::EMERGENCY,
```
is emergency the common default severity? If yes, then it's okay. In WC extensions, most of the time it's just info.

mikejolley replied:
critical and emergency seem pretty similar?

sirreal replied:
@mikejolley Are you considering omitting some log levels? They come from [some](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md) [standards](https://tools.ietf.org/html/rfc5424) definitions, I think the use of these 8 levels is fairly widespread.

mikejolley replied:
O I hadn't seen that. Carry on :p

claudiosanches replied:
The default should be `debug`, since today people only use for debug purposes.

sirreal replied:
I originally chose `emergency` because it was the least invasive default, which log handler implementers can then override with *their* default, and anyone using the logger should manually be able to set.

That said, my original idea is probably not the best. @gedex proposes a sensible default and @claudiosanches proposes log everything by default. For reference, monolog [seems to follow the log everything approach](https://github.com/Seldaek/monolog/blob/4a43d9b17c33bc889674f0fe6c8dc25a2b9b4f73/src/Monolog/Handler/AbstractHandler.php#L30), which seems logical to me. Filtering logs is a conscious decision, not an arbitrary default.

claudiosanches replied:
@sirreal I understand your point and still is just a abstract class and should override inside the other methods.
Not a problem for me.

But just for you have some idea of what people are using it today:

PayPal example:

```
11-10-2016 @ 14:27:08 - Generating payment form for order 450. Notify URL: http://woo.dev/wc-api/WC_Gateway_Paypal/
11-10-2016 @ 14:27:08 - PayPal Request Args for order 450: cmd=_cart&business=XXXXXXXXXx
```

Stripe example:

```
10-20-2016 @ 21:26:31 - Info: Begin processing payment for order 439 for the amount of 37.70
10-20-2016 @ 21:26:31 - charges request: Array
(
    [currency] => brl
    [amount] => 3770
    [description] => WooCommerce Dev - Order 439
    [capture] => true
    [expand[]] => balance_transaction
    [source] => tok_196pvg4p6qOhgF7yA4iFfJPa
)

10-20-2016 @ 21:26:32 - Processing response: stdClass Object
(
    [id] => ch_196pvi4p6qOhgF7ydnQ3ML8N
    [object] => charge
    [amount] => 3770
    ...
    [status] => succeeded
)

10-20-2016 @ 21:26:49 - Success: Stripe charge complete (Charge ID: ch_196pvi4p6qOhgF7ydnQ3ML8N)
```

So you need to be careful if you display a message like "EMERGENCY" or "CRITICAL" by default in from a debug message, since the users can get really concerned about it.

sirreal replied:
@claudiosanches sorry, my response may not have been as clear as it should be. I agree with you! `DEBUG` by default.

sirreal replied:
> So you need to be careful if you display a message like "EMERGENCY" or "CRITICAL" by default in from a debug message, since the users can get really concerned about it.

To clarify, log messages have **no default level**, a user *must* provide a level when they `log` a message. The only way around this is by using the deprecated `WC_Logger->add()`, which I believe I set to log an `info` level message.

The portion of code we're discussing here is the default threshold for `WC_Log_Handler`. `WC_Logger` delegates to work of actually producing some log output to implementations of `WC_Log_Handler`. The handler has a threshold which is the minimum log level it will handle. This way a handler can ignore some messages, for example an email handler probably doesn't want to send `debug` messages, so it would have a higher threshold.

This will not affect any *content* of log messages. Users will never see `EMERGENCY [...]` unless an actually emergency level log has been written.

gedex commented on includes/log-handlers/class-wc-log-handler-file.php:
```diff
@@ -0,0 +1,213 @@
+	 *
+	 * @param string $level emergency|alert|critical|error|warning|notice|info|debug
+	 * @param int $timestamp
+	 * @param string $message
```
Missing param description for `$context`.

gedex replied:
Also need description of each param in docblock.

gedex commented:
@sirreal I took quick look at the code and quick scanning. Overall LGTM, nice works! I haven't tested it yet, but will dedicate some time this week to play with this.

@mikejolley @claudiosan please leave any feedback you have.

sirreal committed:
Fix docblocks

sirreal committed:
Fix scrutinizer issues

sirreal committed:
Change defualt log handler threshold to DEBUG following discussion with team members.

sirreal committed:
Fix file handle regression

sirreal committed:
Add `setMethods` to test mocks

This may improve php 5.2 test failures

sirreal committed:
Add test for log() === [level_short_method]()

sirreal committed:
Defauly log file level is DEBUG

sirreal committed:
Restructure log tests to own directory

Also moves parent::tearDown to end of tearDown
Fix phpcs problems.

sirreal committed:
Fix tests for PHP 5.2

mikejolley commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,135 @@
+	 * @var int 0-8 minimum level severity for handling log entry.
+	 * @access private
+	 */
+	protected $_threshold;
```
_ prefix needed? I know WP used that kind of naming structure a long time ago before it supported private/protected vars, but I don't think it's important anymore.

claudiosanches replied:
Not needed any more and it's already protected, so no need for `@access private`, or if really need to be `private` can change it.

mikejolley commented on includes/log-handlers/class-wc-log-handler-email.php:
```diff
@@ -0,0 +1,118 @@
+	 * @return string subject
+	 */
+	public function get_subject( $level, $timestamp, $message, $context ) {
+		return sprintf( __( '[%s] WooCommerce log message', 'woocommerce' ), $level );
```
How about adding something like the site name here so users who manage multiple stores can clearly see where the log message comes from?

sirreal replied:
The subject and body of log emails both need work. Site name should definitely be included in the subject.

mikejolley commented on includes/log-handlers/class-wc-log-handler-email.php:
```diff
@@ -0,0 +1,118 @@
+	 *
+	 * @return string subject
+	 */
+	public function get_subject( $level, $timestamp, $message, $context ) {
```
If context is optional, set a default.

sirreal replied:
`$context` is optional at the `WC_Logger` level, where there is a default (`array()`). I'd prefer to remove the optional part of the doc block and ensure that a `$context` is being passed around.

sirreal committed:
Remove _ from private/protected properties

sirreal committed:
Remove optional from required log handler context array

claudiosanches commented on includes/log-handlers/class-wc-log-handler-email.php:
```diff
@@ -0,0 +1,118 @@
+	 */
+	public function get_body( $level, $timestamp, $message, $context ) {
+		$entry = $this->format_entry( $level, $timestamp, $message, $context );
+		return sprintf( __( "You have recieved the following WooCommerce log message: \n\n%s", 'woocommerce' ), $entry );
```
Best not use "end of line" inside a translatable string, people some times ignore, not to mention that GlotPress transform it into `↵` (what is really funny since if you click in the "Copy from original" will copy this kind of character too).

See one example with `\n` and `\t`:
![screenshot from 2016-11-16 17-50-51](https://cloud.githubusercontent.com/assets/1264099/20363145/5281f69e-ac25-11e6-8f6d-8dc8156a9ec3.png)

Better change to:

`return sprintf( __( "You have recieved the following WooCommerce log message: %s", 'woocommerce' ), PHP_EOL . PHP_EOL . $entry );`

And better prefer `PHP_EOL` since is more safe for Windows and Unix.

claudiosanches commented on includes/log-handlers/class-wc-log-handler-file.php:
```diff
@@ -0,0 +1,218 @@
+				$handle = 'log';
+			}
+			$message = apply_filters( 'woocommerce_logger_add_message', $message, $handle );
+			$time = date_i18n( 'm-d-Y @ H:i:s -' );
```
This `-` should be in the string `"{$time} - {$message}"`

gedex replied:
Standard date format should be `Y-m-d`, yes? Or it doesn't matter here?

sirreal replied:
I maintained the format here for compatibility with legacy `->add` calls. Calls to the updated `->log`, `->debug`, etc. are formatted with `c` (ISO 8601 date 2004-02-12T15:19:21+00:00). I found that format in a logger spec and it seemed appropriate. I'm open to discuss different alternatives.

gedex replied:
Ok makes sense. I think it's okay for deprecated `add`. Assuming there are some merchants parsing the current log format, we don't want to break it.

The recommended way to use future logger would be to call level-based methods (e.g. `debug`), right?

sirreal replied:
That's most comfortable and easy to read I think. It may also help prevent calling `log` with a bad level string.

claudiosanches replied:
@gedex good point about the date.

@sirreal better apply the standard here, not a problem to change the date format, even more that is more easy to read this new format.

sirreal committed:
Improve log email message format.

Suggestion via https://github.com/woocommerce/woocommerce/pull/12340#pullrequestreview-8895028

sirreal committed:
Include site name in log email message

Suggestion via https://github.com/woocommerce/woocommerce/pull/12340#discussion_r88213040

sirreal committed:
Simplify default log entry format

Remove punctuation. Space separate values:

TIMESTAMP LEVEL MESSAGE

sirreal committed:
Improve logic for legacy `->add` message formatting

Suggestion via https://github.com/woocommerce/woocommerce/pull/12340#discussion_r88318220

sirreal committed:
Restore log deletion from admin log viewer page

WC_Logger no longer controls logging and has no knowledge of files.
Correctly handle `remove_log` action via file log handler.

sirreal committed:
Fix phpcs warning

sirreal committed:
Prevent possible undefined index warning

sirreal committed:
Add specific file log handler tests

sirreal committed:
Rotate logs

WC_Log_Handler_File now includes log rotation.
Add and improve tests.

sirreal committed:
Improve handle signature

The signature of WC_Log_Handler::handle mirrors log messages which is more intuitive.
Method docblocks have also been improved and are more complete.

sirreal committed:
Prefer PHP_EOL over "\n"

sirreal committed:
Add DB log handler

sirreal committed:
Update @package for log handlers

sirreal committed:
Implements database log viewer

sirreal committed:
Log viewer via html template

sirreal committed:
Add level views to db log table

sirreal committed:
Version 1.0.0 for abstract wc log handler

sirreal committed:
Provide correct class name to woocommerce_log_handler_set_threshold filter

sirreal committed:
Add woocommerce_format_log_entry filter to base WC_Log_Handler class

sirreal committed:
Add wc_safe_dump function

Some "safe" server configurations disable certain functions. `wc_safe_dump` attempts to
safely call a series of alternatives suitable for dumping an expression.

sirreal committed:
Fix phpcs issues

sirreal commented:
TODO:

- [ ] Add "flush" action to db log view page to clear all logs.
- [ ] Add a configuration option to select your default log handler (file or db).
- [ ] Update `WC_Logger::add` calls throughout WooCommerce.
- [ ] Update debugging calls (`print_r`) throughout WooCommerce to use `wc_safe_dump`.

sirreal committed:
Update paypal gateway log method

sirreal committed:
update paypal gateway logs

sirreal committed:
remove print_r calls from paypal handler

sirreal committed:
Update logger calls

sirreal committed:
Update print_r to use wc_safe_dump

dnaber-de commented:
Why not just _implement_ `Psr\Log\LoggerInterface` instead of «loosely adhere» to it? That would allow the most flexibility as one could replace the complete logger.

GaryJones commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,141 @@
+		WC_Logger::NOTICE    => 2,
+		WC_Logger::WARNING   => 3,
+		WC_Logger::ERROR     => 4,
+		WC_Logger::CRITICAL  => 6,
```
Where's `5`?

sirreal replied:
Whoops, thanks!

sirreal committed:
Merge branch 'updateLogCalls' into develop

* updateLogCalls:
  Update print_r to use wc_safe_dump
  Update logger calls
  remove print_r calls from paypal handler
  update paypal gateway logs
  Update paypal gateway log method

sirreal committed:
Include missing 5 in log level severities

sirreal committed:
Add severity level tests

sirreal committed:
Add log flushing to db log viewer

sirreal committed:
Mark db handler add method as static

sirreal commented:
In response to @dnaber-de:
> I wonder why the patch »loosely adhere to the PHP Standards Recommendations« instead of implementing them. The benefit would be that one could replace the logger and use for example Monolog, which has a large set of very useful handlers.

I decided to "loosely" adhere rather than "strictly" adhere for 2 reasons:

1. `WC_Logger` currently accepts all logs through the `add` method. This is not part of the spec and I wanted to maintain it for backwards compatibility. I *did* deprecate this method.
2. The following part of the spec did not seem appropriate to me in this context. Rather than throw an exception (potentially crashing mid-transaction), I added a `_doing_it_wrong` message.
    > Calling this method with a level not defined by this specification MUST throw a `Psr\Log\InvalidArgumentException` if the implementation does not know about the level

The rest of the spec essentially describes the nine methods which are respected by the proposed implementation:

```php
log( $level, $message, array $context = array() );
emergency( $message, array $context = array() );
// ...
debug( $message, array $context = array() );
```

sirreal committed:
Add DB handler tests

sirreal committed:
Add confirmation to db log flush

sirreal committed:
Remove bad logger test

Test comparing equality of logger results could randomly fail depending on execution time.
The coverage was duplicated by other tests, test was redundant and has been removed.

sirreal committed:
Fix misnamed test class

sirreal committed:
Rename test file log -> logger

sirreal committed:
Prefer "unit-tests" tag for testing log handler

sirreal committed:
Add test for multiple file log handlers writing to same file.

sirreal committed:
Add basic email logger test

dnaber-de commented:
Thanks for the clarification @sirreal.

nb commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,141 @@
+	 * Decode level string into integer.
+	 *
+	 * @param string $level emergency|alert|critical|error|warning|notice|info|debug
+	 * @return int 0 (debug) - 8 (emergency) or -1 if level is not valid
```
Isn’t the emergency index 7 and not 8?

nb commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,141 @@
+	 * @access private
+	 */
+	protected static $log_levels = array(
+		WC_Logger::DEBUG     => 0,
```
Why is it best for the level → number mapping to live in the abstract class, while the string constants live in `WC_Logger`?

sirreal replied:
I think the log levels themselves should live in their own class. The logger uses them, the log handlers use them, but they don't really belong to either. Conversion to integer could be methods on this loglevel class. This is what PSR3 does: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md#5-psrlogloglevel

nb replied:
A new class sounds good.

nb commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,141 @@
+	 */
+	public function __construct( $args = array() ) {
+
+		$args = wp_parse_args( $args, array(
```
Why do we need an array for named arguments if we have only one argument – `threshold`?

sirreal replied:
This gives flexibility to handler implementations. For example, the email handler also accepts `to`, an email or emails which should receive log notifications.

nb replied:
Couldn’t they add more arguments to their own constructors? How does a single argument here prevents them from allowing arbitrary arguments?

nb commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,141 @@
+	 * @return bool True if the log should be handled.
+	 */
+	public function should_handle( $level ) {
+		return $this->threshold <= $this->get_level_severity( $level );
```
Does the threshold decision need to happen in the handler class? What would be different if we don’t store threshold here at all, but expect that the logger only calls us when needed?

sirreal replied:
Currently, the handler decides whether it should handle the log *and* whether the log should "bubble", i.e. be passed to following handlers. If we move the former to the logger, I think the latter should also be moved.

Conceptually, this means that the logger decides what to do with messages, and handlers are responsible for handling any message they receive. These seem to be more in line with the purposes of each class.

This would also improve a part of the current implementation I'm not happy with. Handlers are responsible for returning `true` to bubble the message or `false` to stop bubbling. This works, but seems opaque and failure-prone.

The change would require passing more information to the logger via the `woocommerce_register_log_handlers` filter, e.g.:

```php
// This is significantly more verbose, but tolerable
$handlers = array(
  array(
    'handler' => new WC_Log_Handler_Email(),
    'threshold' => 'critical',
    'bubble' => false,
  ),
);
```

I don't see any significant disadvantages. I think the change would be an improvement.

nb replied:
I can see why we need a per-handler bubbling (shouldn’t we sort them, though?), why do we want to have a per-handler threshold? What are some use-cases when we have different thresholds for different handlers? If there are, would it make sense to support a global threshold, so that handlers that don’t really care have some sane default?

sirreal replied:
For the record, in another channel of communication we decided the following:
* Bubbling is not particularly useful and can be dropped.
* A global threshold is more useful for handlers, but they remain free to implement a higher threshold internally.
* WC_Logger constructor should accept handlers at construction to make instantiation with particular loggers simpler.

nb commented on includes/class-wc-logger.php:
```diff
@@ -5,157 +5,195 @@
+	public function add( $handle, $message ) {
+		_deprecated_function( 'WC_Logger::add', '2.8', 'WC_Logger::log' );
+		$message = apply_filters( 'woocommerce_logger_add_message', $message, $handle );
+		$this->log( self::INFO, $message, array( 'tag' => $handle, '_legacy' => true ) );
```
Why do you think `INFO` is the most suitable default for existing calls?

sirreal replied:
`INFO` seems the most neutral to me. Anything higher seems to indicate a problem of some sort, and `DEBUG`, the only lower level, seems to indicate the message is trivial. This is just a feeling, I do not have any technical reasons to chose `INFO` or any other level as the default for `add`. 

For reference, here are the [descriptions](https://tools.ietf.org/html/rfc5424):

```
              0       Emergency: system is unusable
              1       Alert: action must be taken immediately
              2       Critical: critical conditions
              3       Error: error conditions
              4       Warning: warning conditions
              5       Notice: normal but significant condition
              6       Informational: informational messages
              7       Debug: debug-level messages
```

nb replied:
👍

nb commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,141 @@
+ * @category       Abstract Class
+ * @author         WooThemes
+ */
+abstract class WC_Log_Handler {
```
Why did you decide on an abstract class vs. an interface?

sirreal replied:
To provide the following functionality by default:

1. Threshold and level logic. This reason may be invalidated based on [your other question](https://github.com/woocommerce/woocommerce/pull/12340#pullrequestreview-11564598).
1. Provide default message formatting. A simple coherent message format is something most handlers should be able to leverage. For example, both file and email handlers take advantage of default formatting.
1. Provide default timestamp formatting according to [spec](https://tools.ietf.org/html/rfc5424#section-6.2.3).

nb replied:
👍  OK, that makes sense, thanks!

nb commented on includes/abstracts/abstract-wc-log-handler.php:
```diff
@@ -0,0 +1,141 @@
+	 * @param string $level emergency|alert|critical|error|warning|notice|info|debug
+	 */
+	public function set_threshold( $level ) {
+		$level = apply_filters( 'woocommerce_log_handler_set_threshold', $level, get_class( $this ) );
```
Would it make sense to have a handler name/ID, instead of using the classname here?

sirreal replied:
Name or ID would likely be more useful because they can refer to a specific instance, whereas the class name is shared. At this point `name` makes sense because there is no requirement for it to be unique. The name would need to be provided at instantiation, the `WC_Log_Handler` abstract class can be modified to accept a `name`, and use `get_class( $this )` as a default.

nb commented on includes/wc-core-functions.php:
```diff
@@ -1408,6 +1408,54 @@ function wc_get_logger() {
 /**
+ * Dump an expression safely.
+ *
+ * Some server environments blacklist some debugging functions. This function provides a safe way to
```
Safe from what?

nb commented on tests/unit-tests/log/logger.php:
```diff
@@ -0,0 +1,221 @@
+	 *
+	 * @return WC_Log_Handler[] array of mocked log handlers
+	 */
+	public function _return_bubble_required_handlers() {
```
If this is a public method, why do we need the prefix?

dwainm commented on includes/wc-core-functions.php:
```diff
@@ -1408,6 +1408,54 @@ function wc_get_logger() {
+		array( 'func' => 'serialize', 'args' => array( $expression ) ),
+	);
+
+	$alternatives = apply_filters( 'woocommerce_safe_sump_alternatives', $alternatives, $expression );
```
should be `woocommerce_safe_dump_alternatives` (sump to dump )

dwainm commented on includes/wc-core-functions.php:
```diff
@@ -1408,6 +1408,54 @@ function wc_get_logger() {
+
+	foreach ( $alternatives as $alternative ) {
+		if ( function_exists( $alternative['func'] ) ) {
+			return call_user_func_array( $alternative['func'], $alternative['args'] );
```
Only the first alternative will be called? Return will exit wc_safe_dump when the first one is found?

sirreal replied:
Yes, the intention it to provide output via the first viable alternative.

nb commented on includes/wc-core-functions.php:
```diff
@@ -1408,6 +1408,54 @@ function wc_get_logger() {
+function wc_safe_dump( $expression ) {
+	$alternatives = array(
+		array( 'func' => 'print_r', 'args' => array( $expression, true ) ),
+		array( 'func' => 'var_dump', 'args' => array( $expression, true ) ),
```
[var_dump](http://php.net/var_dump) doesn’t accept a second argument that controls output/return, it always prints.

sirreal replied:
😱  Thanks for catching that. We can probably just drop `var_dump` as an alternative.

nb commented on includes/admin/class-wc-admin-log-table-list.php:
```diff
@@ -0,0 +1,252 @@
+
+				// Level requires special case to order Emergency -> Debug
+				case 'level':
+					$order_by = 'CASE level '
```
How this affect index usage?

sirreal replied:
It makes the `level` key useless for ordering 😢 .

Solution: store level by an integer representation. The only disadvantages I see are that level strings must be translated on write or print and that the DB becomes more difficult to read for a human or third parties. If the latter is an issue, the level string can be stored in an additional field.

sirreal commented on includes/log-handlers/class-wc-log-handler-email.php:
```diff
@@ -0,0 +1,117 @@
+		if ( $this->should_handle( $level ) ) {
+			$subject = $this->get_subject( $timestamp, $level, $message, $context );
+			$body = $this->get_body( $timestamp, $level, $message, $context );
+			wp_mail( $this->to, $subject, $body );
```
That is probably a good idea, under the right conditions this could send enough mail to be harmful.

Are there existing wp_mail throttling solutions?

A couple of options:

1. 1 email/request: store emails and send a summary email at some final action ([`shutdown`](https://developer.wordpress.org/reference/hooks/shutdown/)?). This avoids multiple emails per request, but is not a perfect throttling solution. Emails could be lost if execution is stopped before the action is reached and a lot of emails could still be sent if traffic is high.
1. 0-1 email/X minutes: Store emails in DB, aggregate and send them periodically. This would be a more robust email throttling solution. `wp_cron` would probably have to be involved and possibly more tables added to the db.

sirreal commented:
Migrating to #12575. Please continue any discussion there.

All commentary here has been dealt with in #12575, any exceptions are noted there.
